pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java

```
@@ -0,0 +1,205 @@
+      } catch (Exception e) {
+        failedParses++;
+        LOGGER.error("Unable to parse server {} response due to an error: ", streamResponse.getKey(), e);
+      }
+    }
+    if (failedParses != 0) {
+      LOGGER.error("Unable to parse server {} / {} response due to an error: ", failedParses, serverURLs.size());
+    }
+
+    LOGGER.debug("Retrieved segment metadata from servers.");
+    return segmentsMetadata;
+  }
+
+  private String generateAggregateSegmentMetadataServerURL(String tableNameWithType, List<String> columns,
+      String endpoint) {
+    try {
+      tableNameWithType = URLEncoder.encode(tableNameWithType, StandardCharsets.UTF_8.name());
+      String paramsStr = "";
+      if (columns != null) {
+        List<String> params = new ArrayList<>(columns.size());
+        for (String column : columns) {
+          params.add(String.format("columns=%s", column));
+        }
+        paramsStr = String.join("&", params);
+      }
+      return String.format("%s/tables/%s/metadata?%s", endpoint, tableNameWithType, paramsStr);
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e.getCause());
+    }
+  }
+
+  private String generateSegmentMetadataServerURL(String tableNameWithType, String segmentName, List<String> columns,
+      String endpoint) {
+    try {
+      tableNameWithType = URLEncoder.encode(tableNameWithType, StandardCharsets.UTF_8.name());
+      segmentName = URLEncoder.encode(segmentName, StandardCharsets.UTF_8.name());
+      String paramsStr = "";
+      if (columns != null) {
+        List<String> params = new ArrayList<>(columns.size());
+        for (String column : columns) {
+          params.add(String.format("columns=%s", column));
+        }
+        paramsStr = String.join("&", params);
+      }
+      return String.format("%s/tables/%s/segments/%s/metadata?%s", endpoint, tableNameWithType, segmentName, paramsStr);
+    } catch (UnsupportedEncodingException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)